backend/common/utils/type_conversion_utils.py

```diff
@@ -17,129 +51,132 @@ def get_dtypes_and_schemas_of_dataframe(dataframe: pd.DataFrame):
     return dtypes_by_column_name, schema_type_hints_by_column_name
 
 
-def get_dtype_of_array(array: pd.Series):
-    return get_dtype_and_schema_of_array(array)[0]
+def get_encoding_dtype_of_array(array: Union[np.ndarray, pd.Series]) -> np.dtype:
+    return _get_type_info(array)[0]
+
+
+def get_schema_type_hint_of_array(array: Union[np.ndarray, pd.Series]) -> dict:
+    return _get_type_info(array)[1]
 
 
-def get_schema_type_hint_of_array(array: pd.Series):
-    return get_dtype_and_schema_of_array(array)[1]
+def get_dtype_and_schema_of_array(array: Union[np.ndarray, pd.Series]) -> Tuple[np.dtype, dict]:
+    """Return tuple (encoding_dtype, schema_type_hint)"""
+    return _get_type_info(array)
 
 
-def get_dtype_and_schema_of_array(array: pd.Series):
-    return (
-        get_dtype_from_dtype(array.dtype, array_values=array),
-        get_schema_type_hint_from_dtype(array.dtype, array_values=array),
-    )
+def get_schema_type_hint_from_dtype(dtype) -> dict:
+    res = _get_type_info_from_dtype(dtype)
+    if res is None:
+        raise TypeError(f"Annotations of type {dtype} are unsupported.")
+    else:
+        return res[1]
 
 
-def get_dtype_from_dtype(dtype, array_values=None):
+def _get_type_info_from_dtype(dtype) -> Union[Tuple[np.dtype, dict], None]:
     """
-    Given a data type, finds the equivalent data type that the array should be encoded as. Notably, this is relevant
-    for 64 bit values which will get downcast to 32 bit.
+    Best-effort to determine encoding type and schema hint from a dtype.
+    If this is not possible, or the type is unsupported, return None.
     """
+    if dtype.kind == "b":
+        return (np.uint8, {"type": "boolean"})
 
-    dtype_name = dtype.name
-    dtype_kind = dtype.kind
+    if dtype.kind == "U":
+        return (np.dtype(str), {"type": "string"})
 
-    if dtype_name == "bool":
-        return np.uint8
-    if dtype_name == "object" and dtype_kind == "O":
-        return str
-    if dtype_name == "category":
-        return get_dtype_from_dtype(dtype.categories.dtype, array_values)
+    if dtype.kind in ["i", "u"]:
+        if np.can_cast(dtype, np.int32):
+            return (np.int32, {"type": "int32"})
 
-    if can_cast_to_int32(dtype, array_values):
-        return np.int32
-    if can_cast_to_float32(dtype, array_values):
-        return np.float32
-    if not can_cast_to_float32(dtype, array_values):
-        return np.float64
+    if dtype.kind == "f":
+        if np.can_cast(dtype, np.float32):
+            return (np.float32, {"type": "float32"})
 
-    raise TypeError(f"Annotations of type {dtype} are unsupported.")
+    return None
 
 
-def get_schema_type_hint_from_dtype(dtype, array_values=None):
+def _get_type_info(array: Union[np.ndarray, pd.Series]) -> Tuple[np.dtype, dict]:
```

Comment:
```suggestion
def _get_type_info(array: Union[np.ndarray, pd.Series, pd.Index]) -> Tuple[np.dtype, dict]:
```

Based on the type check on line 103.

Reply:
Ditto for other the other `can_cast` functions.